subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ResolveBuildCacheKeyExecuter.java

```diff
@@ -114,15 +170,21 @@ public void run(BuildOperationContext buildOperationContext) {
         });
     }
 
-    private TaskOutputCachingBuildCacheKey doResolve(TaskInternal task, TaskExecutionContext context) {
-        TaskArtifactState taskState = context.getTaskArtifactState();
-        TaskProperties taskProperties = context.getTaskProperties();
-        TaskOutputCachingBuildCacheKey cacheKey = taskState.calculateCacheKey(context.getAfterPreviousExecution(), taskProperties);
-        if (taskProperties.hasDeclaredOutputs() && cacheKey.isValid()) { // A task with no outputs has no cache key.
-            LogLevel logLevel = buildCacheDebugLogging ? LogLevel.LIFECYCLE : LogLevel.INFO;
-            LOGGER.log(logLevel, "Build cache key for {} is {}", task, cacheKey.getHashCode());
-        }
-        return cacheKey;
+    private TaskOutputCachingBuildCacheKey doResolve(final TaskInternal task, TaskExecutionContext context) {
+        final TaskProperties taskProperties = context.getTaskProperties();
+        return context.getBeforeExecutionState()
+            .map(new java.util.function.Function<BeforeExecutionState, TaskOutputCachingBuildCacheKey>() {
```

Comment:
Ah, right, I wanted to do that actually.

Reply:
OK, I checked it, and the problem is that you need to use Guava's `Function`s with their collections API. We could use the Java 8 stream API, and we should, but let's do this in a separate step at some point.

Reply:
BTW, one of the goals of this exercise is that we can move all this to `:execution`, which would allow us to use lambdas. That would at least hide the ugliness here, and it would also make it easier to switch to the stream API.

Reply:
Since the usages of Guava's `Function` are now gone from this class, we can use an import already.

Reply:
Done.